single_kernel_mongo/managers/ldap.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The managers for the ldap relations between glauth-k8s and mongo charms."""
+
+from __future__ import annotations
+
+from logging import getLogger
+from typing import TYPE_CHECKING
+
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, Relation, StatusBase
+
+from single_kernel_mongo.config.literals import (
+    Substrates,
+    TrustStoreFiles,
+)
+from single_kernel_mongo.config.relations import ExternalRequirerRelations
+from single_kernel_mongo.core.status_provider import StatusProvider
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.exceptions import (
+    DeferrableFailedHookChecksError,
+    LDAPSNotEnabledError,
+    NonDeferrableFailedHookChecksError,
+    WaitingForLdapDataError,
+)
+from single_kernel_mongo.lib.charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateTransferRequires,
+)
+from single_kernel_mongo.lib.charms.glauth_k8s.v0.ldap import LdapRequirer
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.core.operator import OperatorProtocol
+
+logger = getLogger(__name__)
+
+
+class LDAPManager(Object, StatusProvider):
+    """Manage the relation between glauth-k8s and mongos router."""
+
+    def __init__(
+        self,
+        dependent: OperatorProtocol,
+        state: CharmState,
+        substrate: Substrates,
+        relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP,
+        cert_relation_name: ExternalRequirerRelations = ExternalRequirerRelations.LDAP_CERT,
+    ):
+        super().__init__(parent=dependent, key=relation_name)
+        self.dependent = dependent
+        self.charm = dependent.charm
+        self.workload = self.dependent.workload
+        self.state = state
+        self.substrate = substrate
+        self.relation_name = relation_name
+        self.cert_relation_name = cert_relation_name
+        self.ldap_requirer = LdapRequirer(self.charm, self.relation_name)
+        self.certificate_transfer = CertificateTransferRequires(self.charm, self.cert_relation_name)
+
+    def assert_pass_hook_checks(self) -> None:
+        """Runs some hook checks before allowing the hook to run."""
+        if not self.state.db_initialised:
+            raise DeferrableFailedHookChecksError("DB is not initialised")
+        if self.state.is_role(MongoDBRoles.SHARD):
+            self.charm.status_manager.to_blocked("Cannot integrate LDAP with shard.")
```

Comment:
nit: mind doing `set_and_share_status`, makes it easier in my status work when I create constants :)